### PR TITLE
IntervalV2: Use maxDataPoints to calculate correct interval

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:
Part of https://github.com/grafana/grafana/issues/37784

https://github.com/grafana/grafana/pull/37971/files was becoming too big of a PR so I am going to split it into smaller and more focused ones. To migrate prometheus queries to run trough backend, we need to calculate correct interval. Currently, we were using default/hard coded resolution. This PR changes it, so actual maxDataPoints are used to calculate interval. If maxDataPoints are 0, we use the default resolution as before. 

I updated only v2 interval as we are planning to remove interval and it is not used in prometheus/loki/es data sources.

I have also added test cases for this new change.

### CODE REVIEW

File: pkg/tsdb/intervalv2/intervalv2.go [hunk 1]
```diff
@@ -53,16 +53,21 @@ func (i *Interval) Milliseconds() int64 {
 	return i.Value.Nanoseconds() / int64(time.Millisecond)
 }
 
-func (ic *intervalCalculator) Calculate(timerange backend.TimeRange, minInterval time.Duration) Interval {
+func (ic *intervalCalculator) Calculate(timerange backend.TimeRange, minInterval time.Duration, maxDataPoints int64) Interval {
 	to := timerange.To.UnixNano()
 	from := timerange.From.UnixNano()
-	calculatedIntrvl := time.Duration((to - from) / defaultRes)
+	resolution := maxDataPoints
+	if resolution == 0 {
```
Comment (gabor) [inline pkg/tsdb/intervalv2/intervalv2.go:60]: in theory this should never happen anymore, all paths that lead here should calculate maxDataPoints correctly.

but i understand why the check is here for now... i wonder if we should put here some logging perhaps or something that notifies us that we reached a place that should not happen and fixing is needed 😄

**Review by gabor — APPROVED** [review]
code looks good! works fine with prometheus.

File: pkg/tsdb/prometheus/prometheus.go [hunk 2]
```diff
@@ -248,7 +248,7 @@ func (s *Service) parseQuery(queries []backend.DataQuery, dsInfo *DatasourceInfo
 			return nil, err
 		}
 
-		calculatedInterval := s.intervalCalculator.Calculate(queries[0].TimeRange, minInterval)
+		calculatedInterval := s.intervalCalculator.Calculate(queries[0].TimeRange, minInterval, queryModel.MaxDataPoints)
```
Comment (gabor) [inline pkg/tsdb/prometheus/prometheus.go:251]: NOTE: 
the value we are calculating here should be calculated already in `queries[0].Interval` ... but i understand that you want to go with small pull-request-steps :+1:

Comment (gabor) [reply pkg/tsdb/prometheus/prometheus.go:251]: sorry, my mistake here, please ignore this comment.

the value in `queries[0].Interval` did not "respect" `prometheus_min_step`, so it is not the correct value.

**Review by dsotirakis — APPROVED** [review]
Please have a look at Gabor's comments, LGTM otherwise! 😁

File: pkg/tsdb/intervalv2/intervalv2.go (hunk 1 above)
Comment (ivanahuckova) [reply pkg/tsdb/intervalv2/intervalv2.go:60]: @gabor we can log this, but for us to make this log helpful, we would need to know what query resulted in `maxDataPoinits == 0`. Which would. end up with adding more arguments/passiing plog as in `customQueryParametersMiddleware(plog)`. Or we could check it in each data source beforehand?  Do you have any thoughts on which one should be added?

Comment (ivanahuckova) [reply pkg/tsdb/intervalv2/intervalv2.go:60]: @gabor I am going to merge this so I can move on with Prometheus backend refactor, but I added this as a task in https://github.com/grafana/grafana/issues/37784. We can have a chat later about how to notify us about `maxDataPoinits == 0`.

**Pull request merged**